### PR TITLE
Add KafkaOpsRequest API

### PR DESCRIPTION
**Pull request by obaydullahmhs**
Signed-off-by: obaydullahmhs <obaydullah@appscode.com>

Fixes kubedb/project#915

### CODE REVIEW

File: apis/ops/v1alpha1/kafka_ops_helpers.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"kmodules.xyz/client-go/apiextensions"
+)
+
+func (_ KafkaOpsRequest) CustomResourceDefinition() *apiextensions.CustomResourceDefinition {
+	return crds.MustCustomResourceDefinition(SchemeGroupVersion.WithResource(ResourcePluralKafkaOpsRequest))
+}
+
+var _ apis.ResourceInfo = &KafkaOpsRequest{}
```
Comment (raihankhan) [inline apis/ops/v1alpha1/kafka_ops_helpers.go:34]: Move it to top

File: apis/ops/v1alpha1/kafka_ops_helpers.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+func (k KafkaOpsRequest) ResourceSingular() string {
+	return ResourceSingularKafkaOpsRequest
+}
+
+func (k KafkaOpsRequest) ResourcePlural() string {
+	return ResourcePluralKafkaOpsRequest
+}
+
+func (k KafkaOpsRequest) ValidateSpecs() error {
```
Comment (raihankhan) [inline apis/ops/v1alpha1/kafka_ops_helpers.go:56]: Set the receivers to pointer type

File: apis/ops/v1alpha1/kafka_ops_helpers.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+	return k.Spec.DatabaseRef.Name
+}
+
+func (k *KafkaOpsRequest) GetStatus() OpsRequestStatus {
+	return k.Status
+}
+
+func (k *KafkaOpsRequest) SetStatus(s OpsRequestStatus) {
+	k.Status = s
+}
```
Comment (raihankhan) [inline apis/ops/v1alpha1/kafka_ops_helpers.go:84]: Remove these methods if not required

File: apis/ops/v1alpha1/kafka_ops_types.go [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=kafkaopsrequests,singular=kafkaopsrequest,shortName=kfops,categories={datastore,kubedb,appscode}
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name="Type",type="string",JSONPath=".spec.type"
+// +kubebuilder:printcolumn:name="Status",type="string",JSONPath=".status.phase"
+// +kubebuilder:printcolumn:name="Age",type="date",JSONPath=".metadata.creationTimestamp"
+type KafkaOpsRequest struct {
+	metav1.TypeMeta   `json:",omitempty"`
```
Comment (raihankhan) [inline apis/ops/v1alpha1/kafka_ops_types.go:46]: use 
```
 metav1.TypeMeta   `json:",inline,omitempty"` 
 ```

File: apis/ops/v1alpha1/kafka_ops_types.go [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+	// Specifies information necessary for vertical scaling
+	VerticalScaling *KafkaVerticalScalingSpec `json:"verticalScaling,omitempty"`
+	// Specifies information necessary for volume expansion
+	VolumeExpansion *KafkaVolumeExpansionSpec `json:"volumeExpansion,omitempty"`
+	// Specifies information necessary for custom configuration of Kafka
+	Configuration *KafkaCustomConfigurationSpec `json:"configuration,omitempty"`
+	// Specifies information necessary for configuring TLS
+	TLS *TLSSpec `json:"tls,omitempty"`
+	// Specifies information necessary for restarting database
+	Restart *RestartSpec `json:"restart,omitempty"`
```
Comment (raihankhan) [inline apis/ops/v1alpha1/kafka_ops_types.go:71]: Also add reprovision opsRequest

File: apis/ops/v1alpha1/kafka_ops_types.go [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+type KafkaCustomConfigurationSpec struct {
+	// ConfigSecret is an optional field to provide custom configuration file for database.
+	// +optional
+	ConfigSecret *core.LocalObjectReference `json:"configSecret,omitempty"`
+	// ApplyConfig is an optional field to provide Kafka configuration.
+	// Provided configuration will be applied to config files stored in ConfigSecret.
+	// If the ConfigSecret is missing, the operator will create a new k8s secret by the
+	// following naming convention: {db-name}-user-config .
+	// Expected input format:
+	//	applyConfig:
```
Comment (raihankhan) [inline apis/ops/v1alpha1/kafka_ops_types.go:154]: Make sure extra configs are mounted from operator

**Review by raihankhan — CHANGES_REQUESTED** [review]

File: apis/ops/v1alpha1/kafka_ops_helpers.go (hunk 3 above)
Comment (obaydullahmhs) [reply apis/ops/v1alpha1/kafka_ops_helpers.go:84]: As it is interface implementation, these methods should be implemented.

File: apis/ops/v1alpha1/kafka_ops_types.go [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+	// Number of controller nodes
+	Controller *int32 `json:"controller,omitempty"`
+}
+
+// KafkaVerticalScalingSpec contains the vertical scaling information of a Kafka cluster
+type KafkaVerticalScalingSpec struct {
+	// Resource spec for combined nodes
+	Node *core.ResourceRequirements `json:"node,omitempty"`
+	// Resource spec for exporter sidecar
+	Exporter *core.ResourceRequirements `json:"exporter,omitempty"`
```
Comment (raihankhan) [inline apis/ops/v1alpha1/kafka_ops_types.go:114]: Remove it as Kafka don't have exporter sidecar.

**Review by raihankhan — CHANGES_REQUESTED** [review]

**Pull request merged**